### PR TITLE
Increment 13

### PR DESCRIPTION
**Pull request by Taboon**

### CODE REVIEW

File: internal/storage/postgre.go [hunk 1]
```diff
@@ -23,32 +25,181 @@ func NewPostgreBase(db *pgx.Conn, log *zap.Logger) *Postgre {
...
+	ctx, cancel := context.WithTimeout(context.Background(), 1*time.Second)
+	defer cancel()
+	_, err := p.db.Exec(ctx, "INSERT INTO urls (id, url) VALUES ($1, $2)", data.ID, data.URL)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (p *Postgre) AddBatchURL(urls map[string]ReqBatchJSON) error {
```
Comment (arttet) [inline internal/storage/postgre.go]: CHANGE REQUEST: you should provide `context`

File: internal/storage/postgre.go [hunk 2]
```diff
@@ -23,32 +25,181 @@ func NewPostgreBase(db *pgx.Conn, log *zap.Logger) *Postgre {
...
-	p.Log.Debug("Есть соединение с БД")
+	p.Log.Info("Есть соединение с БД")
 	return nil
 }
 
 func (p *Postgre) AddURL(data URLData) error {
-	//TODO implement me
-	panic("implement me1")
+	p.Log.Debug("Добавляем URL в базу данных", zap.String("url", data.URL))
+	ctx, cancel := context.WithTimeout(context.Background(), 1*time.Second)
```
Comment (arttet) [inline internal/storage/postgre.go]: CHANGE REQUEST: you should use `context` from handlers.

File: internal/storage/postgre.go [hunk 3]
```diff
@@ -23,32 +25,181 @@ func NewPostgreBase(db *pgx.Conn, log *zap.Logger) *Postgre {
...
+		if !errors.Is(err, pgx.ErrNoRows) {
+			p.Log.Error("Другая ошибка запроса", zap.Error(err))
+			return URLData{}, false, err
+		}
+	}
+	p.Log.Debug("Возвращаем URLData", zap.String("url", u), zap.String("id", i))
+	return URLData{URL: u, ID: i}, true, nil
+}
+
+func (p *Postgre) CheckBatchURL(urls *[]ReqBatchJSON) (*[]ReqBatchJSON, error) {
```
Comment (arttet) [inline internal/storage/postgre.go]: CHANGE REQUEST: please find another solution.

**Review by arttet — CHANGES_REQUESTED** [review]

File: internal/storage/postgre.go (hunk 1 above)
Comment (Taboon) [reply internal/storage/postgre.go]: done

File: internal/storage/postgre.go (hunk 2 above)
Comment (Taboon) [reply internal/storage/postgre.go]: done

File: internal/storage/postgre.go (hunk 3 above)
Comment (Taboon) [reply internal/storage/postgre.go]: done

File: internal/db/migrations/20240420094419_urlshortner_migration.sql [hunk 4]
```diff
@@ -0,0 +1,13 @@
+-- +goose Up
+-- +goose StatementBegin
+CREATE TABLE IF NOT EXISTS urls (
```
Comment (arttet) [inline internal/db/migrations/20240420094419_urlshortner_migration.sql]:
```suggestion
CREATE TABLE urls (
```

File: internal/db/migrations/20240420094419_urlshortner_migration.sql [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+                                    id TEXT PRIMARY KEY,
+                                    url TEXT,
+                                    count INT
```
Comment (arttet) [inline internal/db/migrations/20240420094419_urlshortner_migration.sql]: CLARIFICATION: what is it?

File: internal/db/migrations/20240420094419_urlshortner_migration.sql [hunk 6]
```diff
@@ -0,0 +1,13 @@
...
+);
+-- +goose StatementEnd
+
+-- +goose Down
+-- +goose StatementBegin
+SELECT 'down SQL query';
```
Comment (arttet) [inline internal/db/migrations/20240420094419_urlshortner_migration.sql]: CLARIFICATION: what is it?

File: internal/domain/usecase/shortURL.go [hunk 7]
```diff
@@ -16,46 +18,141 @@ const (
...
 		if err != nil {
 			u.Log.Error("Ошибка сохранения бекапа")
 		}
 	}
+
 	return id, nil
 }
 
+func hasDuplicates(urls *[]storage.ReqBatchJSON) {
+	urlMap := make(map[string]bool)
```
Comment (arttet) [inline internal/domain/usecase/shortURL.go]:
```suggestion
	urlMap := make(map[string]bool, 0, len(urls))
```

File: internal/domain/usecase/shortURL.go [hunk 8]
```diff
@@ -16,46 +18,141 @@ const (
...
+
+	for i, item := range *urls {
+		if _, ok := urlMap[item.URL]; ok {
+			(*urls)[i].Exist = true // Дубликат найден
+		}
+		urlMap[item.URL] = true
+	}
+}
+
+func (u *URLProcessor) BatchURLSaver(ctx context.Context, urls *[]storage.ReqBatchJSON) (map[string]storage.ReqBatchJSON, error) {
```
Comment (arttet) [inline internal/domain/usecase/shortURL.go]: CHANGE REQUEST: please simplify this function

File: internal/domain/usecase/shortURL.go [hunk 9]
```diff
@@ -16,46 +18,141 @@ const (
...
+		urlsToDB[i] = v
+	}
+
+	return urlsToDB, nil
+}
+
 func (u *URLProcessor) generateID() string {
 	u.Log.Debug("Генерируем ID")
+
+	ctx, cancel := context.WithCancel(context.Background())
```
Comment (arttet) [inline internal/domain/usecase/shortURL.go]: CHANGE REQUEST: it should be a parameter of this function

File: internal/server/handlers.go [hunk 10]
```diff
@@ -59,15 +65,20 @@ func (s *Server) getURL(w http.ResponseWriter, r *http.Request) {
 		return
 	}
 
-	id, err := s.P.URLSaver(url)
+	w.Header().Set("Content-Type", "text/plain")
+
+	id, err := s.P.URLSaver(r.Context(), url)
 	if err != nil {
```
Comment (arttet) [inline internal/server/handlers.go]:
```suggestion
	if err != nil {
		if errors.Is(err, entity.ErrURLExist) {
			w.WriteHeader(http.StatusConflict)

			return
		} else {
			http.Error(w, "Не удалось сохранить URL: "+err.Error(), http.StatusBadRequest)
			return
		}
  w.WriteHeader(http.StatusCreated)
```

File: internal/server/handlers.go [hunk 11]
```diff
@@ -124,6 +143,71 @@ func (s *Server) shortenJSON(w http.ResponseWriter, r *http.Request) {
 		return
 	}
 
+	_, err = w.Write(resp)
+
+	if err != nil {
+		http.Error(w, "Не удалось записать ответ: "+err.Error(), http.StatusBadRequest)
```
Comment (arttet) [inline internal/server/handlers.go]: CHANGE REQUEST: it should be the 500 error

File: internal/server/handlers.go [hunk 12]
```diff
@@ -124,6 +143,71 @@ func (s *Server) shortenJSON(w http.ResponseWriter, r *http.Request) {
...
+		return
+	}
+}
+
+func (s *Server) shortenBatchJSON(w http.ResponseWriter, r *http.Request) {
```
Comment (arttet) [inline internal/server/handlers.go]: CHANGE REQUEST: please enable the following linter
```
  funlen:
    # Checks the number of lines in a function.
    # If lower than 0, disable the check.
    # Default: 60
    lines: 32
```

File: internal/server/server.go [hunk 13]
```diff
@@ -33,5 +33,6 @@ func (s *Server) URLRouter() chi.Router {
 	r.Get("/ping", s.Log.RequestLogger(s.ping))
```
Comment (arttet) [inline internal/server/server.go:44]: CHANGE REQUEST: please revert this changes from `main`

File: internal/storage/file_storage.go [hunk 14]
```diff
@@ -15,6 +16,31 @@ type FileStorage struct {
 	Log      *zap.Logger
 }
 
+var urlData = URLData{}
```
Comment (arttet) [inline internal/storage/file_storage.go]: CHANGE REQUEST: please don't use global variables

File: internal/storage/postgre.go [hunk 15]
```diff
@@ -23,35 +27,161 @@ func NewPostgreBase(db *pgx.Conn, log *zap.Logger) *Postgre {
...
+	db, err := sql.Open("pgx", dsn)
+	if err != nil {
+		return err
+	}
+	defer db.Close()
+
+	if err := goose.SetDialect("postgres"); err != nil {
+		return err
+	}
+	if err := goose.Up(db, "internal/db/migrations"); err != nil {
```
Comment (arttet) [inline internal/storage/postgre.go]: CHANGE REQUEST: this is a magic string

File: internal/storage/postgre.go [hunk 16]
```diff
@@ -23,35 +27,161 @@ func NewPostgreBase(db *pgx.Conn, log *zap.Logger) *Postgre {
...
+func (p *Postgre) AddBatchURL(ctx context.Context, urls map[string]ReqBatchJSON) error {
+	tx, err := p.db.Begin(ctx)
+	if err != nil {
+		return err
+	}
+	for id, v := range urls {
+		// все изменения записываются в транзакцию
+		if v.Valid && !v.Exist {
+			_, err := tx.Exec(ctx,
+				"INSERT INTO urls (id, url)"+
```
Comment (arttet) [inline internal/storage/postgre.go]:
```suggestion
				`INSERT INTO urls (id, url) VALUES($1, $2)`
```

File: internal/storage/postgre.go [hunk 17]
```diff
@@ -23,35 +27,161 @@ func NewPostgreBase(db *pgx.Conn, log *zap.Logger) *Postgre {
...
-	panic("implement me2")
+func (p *Postgre) AddBatchURL(ctx context.Context, urls map[string]ReqBatchJSON) error {
+	tx, err := p.db.Begin(ctx)
+	if err != nil {
+		return err
+	}
+	for id, v := range urls {
+		// все изменения записываются в транзакцию
+		if v.Valid && !v.Exist {
+			_, err := tx.Exec(ctx,
```
Comment (arttet) [inline internal/storage/postgre.go]: CHANGE REQUEST: please optimize it

https://stackoverflow.com/questions/70489530/golang-route-for-insert-preparecontext-error

File: internal/storage/postgre.go [hunk 18]
```diff
@@ -23,35 +27,161 @@ func NewPostgreBase(db *pgx.Conn, log *zap.Logger) *Postgre {
...
+	return URLData{URL: u, ID: i}, true, nil
+}
+
+func (p *Postgre) CheckURL(ctx context.Context, url string) (URLData, bool, error) {
+	var i string
+	var u string
+
+	p.Log.Debug("Проверяем URL в базе данных", zap.String("url", url))
+
+	err := p.db.QueryRow(ctx, "SELECT id, url FROM urls WHERE url = $1", url).Scan(&i, &u)
```
Comment (arttet) [inline internal/storage/postgre.go]: CHANGE REQUEST: please build SQL execution plan query cost.

File: internal/storage/postgre.go [hunk 19]
```diff
@@ -23,35 +27,161 @@ func NewPostgreBase(db *pgx.Conn, log *zap.Logger) *Postgre {
...
+	var queryInsert string
+	for i := 1; i <= len(values); i++ {
+		queryInsert += fmt.Sprintf("$%v", i)
+		if i < len(values) {
+			queryInsert += ","
+		}
+	}
+
+	// Проверка существования урлов в базе данных
+	query := "SELECT url FROM urls WHERE url IN (" + queryInsert + ")"
```
Comment (arttet) [inline internal/storage/postgre.go]: CHANGE REQUEST: you should use ` INSERT ... ON CONFLICT `

**Review by arttet — CHANGES_REQUESTED** [review]

File: internal/db/migrations/20240420094419_urlshortner_migration.sql (hunk 4 above)
Comment (Taboon) [reply internal/db/migrations/20240420094419_urlshortner_migration.sql]: done

File: internal/db/migrations/20240420094419_urlshortner_migration.sql (hunk 6 above)
Comment (Taboon) [reply internal/db/migrations/20240420094419_urlshortner_migration.sql]: unused code has been removed

File: internal/db/migrations/20240420094419_urlshortner_migration.sql (hunk 5 above)
Comment (Taboon) [reply internal/db/migrations/20240420094419_urlshortner_migration.sql]: removed

File: internal/domain/usecase/shortURL.go (hunk 7 above)
Comment (Taboon) [reply internal/domain/usecase/shortURL.go]: done

File: internal/db/migrations/20240420094419_urlshortner_migration.sql [hunk 20]
```diff
@@ -0,0 +1,7 @@
+-- +goose Up
+-- +goose StatementBegin
+CREATE TABLE urls (
+                                    id TEXT PRIMARY KEY,
```
Comment (arttet) [inline internal/db/migrations/20240420094419_urlshortner_migration.sql:4]:
```suggestion
  short_url VARCHAR(8) PRIMARY KEY,
```

File: internal/db/migrations/20240420094419_urlshortner_migration.sql [hunk 21]
```diff
@@ -0,0 +1,7 @@
...
+                                    url TEXT UNIQUE
```
Comment (arttet) [inline internal/db/migrations/20240420094419_urlshortner_migration.sql:5]:
```suggestion 
  original_url VARCHAR(2048) UNIQUE
```

File: internal/db/migrations/20240420094419_urlshortner_migration.sql [hunk 22]
```diff
@@ -0,0 +1,7 @@
+-- +goose Up
+-- +goose StatementBegin
+CREATE TABLE urls (
```
Comment (arttet) [inline internal/db/migrations/20240420094419_urlshortner_migration.sql:3]:
```suggestion
CREATE TABLE url (
```

File: internal/domain/usecase/getURL.go [hunk 23]
```diff
@@ -1,12 +1,13 @@
 package usecase
 
 import (
+	"context"
```
Comment (arttet) [inline internal/domain/usecase/getURL.go:4]: CHANGE REQUEST: please sort imports

File: internal/server/handlers_test.go [hunk 24]
```diff
@@ -89,25 +86,19 @@ func TestSendUrl(t *testing.T) {
...
 			fmt.Println(string(body))
 			fmt.Println(resp)
 
 			defer resp.Body.Close()
 
-			id, _, err := s.P.Repo.CheckID("AAAAaaaa")
-			if err != nil {
-				return
-			}
+			id, _, err := s.P.Repo.CheckID(context.Background(), "AAAAaaaa")
```
Comment (arttet) [inline internal/server/handlers_test.go:100]: No comment

File: internal/storage/postgre.go [hunk 25]
```diff
@@ -16,42 +24,195 @@ type Postgre struct {
...
+
+	for _, v := range *b {
+		// если данные не валидны, пропускаем текущую итерацию
+		if v.Err != nil {
+			continue
+		}
+
+		p.Log.Debug("Пытаемся добавить URL в БД", zap.String("url", v.URL), zap.String("id", v.ID))
+
+		_, err := tx.Exec(ctx, "INSERT INTO urls (id, url) VALUES ($1, $2)", v.ID, v.URL)
```
Comment (arttet) [inline internal/storage/postgre.go:90]: It would be better if you use prepared statements

https://medium.com/@xuan11290/understand-mysql-prepared-statement-1eb1bda59f7b

File: internal/storage/postgre.go [hunk 26]
```diff
@@ -16,42 +24,195 @@ type Postgre struct {
 
 var _ Repository = (*Postgre)(nil)
 
+const dirMigrations = "internal/db/migrations"
```
Comment (arttet) [inline internal/storage/postgre.go:27]: CHANGE REQUEST: this is a magic string. It won't work if this path is changed.

Comment (arttet) [inline internal/storage/postgre.go:27]: CHANGE REQUEST: please move migrations in the root dir.

File: internal/storage/repository.go [hunk 27]
```diff
@@ -22,3 +26,19 @@ type URLData struct {
 	URL string
 	ID  string
 }
+
+type ReqBatchURLs []ReqBatchURL
```
Comment (arttet) [inline internal/storage/repository.go:30]: CHANGE REQUEST: it should be in the `model` or `dto` packages.

**Review by arttet — APPROVED** [review]

File: internal/db/migrations/20240420094419_urlshortner_migration.sql (hunk 20 above)
Comment (Taboon) [reply internal/db/migrations/20240420094419_urlshortner_migration.sql:4]: done

File: internal/db/migrations/20240420094419_urlshortner_migration.sql (hunk 22 above)
Comment (Taboon) [reply internal/db/migrations/20240420094419_urlshortner_migration.sql:3]: done

File: internal/storage/repository.go (hunk 27 above)
Comment (Taboon) [reply internal/storage/repository.go:30]: done
